### ISSUE TITLE
Training a keras application doesn't work with tensorflow backend, but it does work with pytorch.

### ISSUE THREAD

**Issue by odinsbane** [keras-team/keras#19061]
When I train a model built with a keras.applications app using a tensorflow backend, it never finishes a batch. When I use pytorch as a back end it trains fine.

Here is a working example:

```
    mdl = keras.applications.MobileNet()
    op = mdl.layers[85]
    op2 = keras.layers.Conv2DTranspose(1, (32, 32), (32, 32))(op.output)
    model2 = keras.models.Model(inputs = mdl.inputs, outputs = op2)

    for layer in model2.layers:
        if layer.name == "train_me":
            print("training")
        else:
            layer.trainable = False
    
    x = numpy.random.random( (4, 224, 224, 3))
    y = numpy.random.random( (4, 224, 224, 1))

    model2.compile(optimizer = keras.optimizers.Adam(0.0001), loss="mean_squared_error")

    model2.fit(x, y)
```

If I run this with a tensorflow backend, then it never finishes. If I run it with a pytorch backend then it finishes very quickly, less than 1 second. I haven't seen the tensorflow version finish yet.

This is a warning I get from tensorflow:

>2024-01-16 10:44:20.910445: E external/local_xla/xla/service/slow_operation_alarm.cc:65] Trying algorithm eng27{k2=0,k12=-1,k13=2,k14=3,k15=0,k17=171,k18=1,k23=0} for conv (f32[1,4,224,224]{3,2,1,0}, u8[0]{0}) custom-call(f32[1,1024,32,32]{3,2,1,0}, f32[4,1024,193,193]{3,2,1,0}, f32[4]{0}), window={size=193x193 pad=192_192x192_192 rhs_reversal=1x1}, dim_labels=bf01_oi01->bf01, custom_call_target="__cudnn$convBiasActivationForward", backend_config={"conv_result_scale":1,"activation_mode":"kNone","side_input_scale":0,"leakyrelu_alpha":0} is taking a while...

**Comment by odinsbane** [comment]
Actually, this simple example did finally finish with tensorflow, It took 9 minutes. If I hide the gpu by `export CUDA_VISIBLE_DEVICES=""` then it finishes in 5 seconds.

**Comment by sachinprasadhs** [comment]
I was able to run the code successfully using TensorFlow within few seconds, here is the Gist attached for reference https://gist.github.com/sachinprasadhs/d8667509dd1ad6d22d88336eab821a0f

**Comment by odinsbane** [comment]
In the notebook you provided it says, the model has no trainable weights.
When I ran it, the last layer was trainable. Does that make a difference?

**Comment by sachinprasadhs** [comment]
In your code, you have set `trainable=False` and there is no layer with the name "train_me", so it would have 0 trainable parameters.

```
        if layer.name == "train_me":
            print("training")
        else:
            layer.trainable = False
```

```python
 Total params: 4,277,441 (16.32 MB)
 Trainable params: 0 (0.00 B)
 Non-trainable params: 4,277,441 (16.32 MB)
```

**Comment by odinsbane** [comment]
The layer I add, I give it the name train_me so it has trainable weights. I
am not sure how that got removed from the example.

**Comment by odinsbane** [comment]
I checked again, even when there are no trainable weights this takes a long time on my computer with a tensorflow backend. It is the same on both wsl2 and linux for me. 

Is the notebook your creating using the gpu?

**Comment by odinsbane** [comment]
A little more debugging info. If I use keras 2.15 that was installed with tensorflow then it works as expected. The main difference I see is errors from ptx.

>Unsupported .version 7.8; current version is '7.5'  
ptxas fatal   : Ptx assembly aborted due to errors

It seems like my cuda/tensorflow drivers might not be correct for keras 3. I installed them using the recommended instructions from keras website. Eg. `pip install tensorflow[and-cuda]` then `pip install --upgrade keras`.

**Comment by dhantule** [comment]
Hi @odinsbane, thanks for reporting. Are you still able to reproduce this ? To work with GPU with different backends you check the dependency versions used by colab [here ](https://colab.sandbox.google.com/drive/13cpd3wCwEHpsmypY9o6XB6rXgBm5oSxu). You can read more about GPU dependencies [here](https://keras.io/getting_started/#gpu-dependencies).

**Comment by odinsbane** [comment]
I tried it with keras==3.5.0 and  tensorflow==2.17.0 and it still takes an incredibly long time with the gpu. The gpu is working at 100% after 3 minutes it says that an operation is slow. If I disable the GPU by setting `CUDA_VISIBLE_DEVICES=""` then it finishes quickly. So nothing has changed.

**Comment by odinsbane** [comment]
I tried it on windows wsl, and using a linux machine. They both have the same problem. It did finally finish

>I0000 00:00:1737102874.929431  490873 device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.
1/1 ━━━━━━━━━━━━━━━━━━━━ 997s 997s/step - loss: 0.3328

**Comment by LakshmiKalaKadali** [comment]
Hi @odinsbane,

I was able to reproduce the hang you experienced with the TensorFlow backend. The reported behavior is due to  two things.
1. The XLA/JIT Compilation hang is due to large kernel and stride (32, 32) in your Conv2DTranspose layer causes TensorFlow's XLA compiler to take an extremely long time to autotune/compile the graph on the first batch. This makes the model appear to be hung.
2.  In your original code, the loop to unfreeze the layer was likely failing to find the target layer because it wasn't explicitly named, resulting in a model with 0 trainable parameters.

Please disable JIT(Set jit_compile=False) in model.compile(). This bypasses the expensive XLA compilation phase for this specific architecture and name your layer and ensure it is set to trainable=True. Could you please verify this [colab notebook](https://colab.sandbox.google.com/gist/LakshmiKalaKadali/fe9c821ad3e8f5e420326c63bc60e3b8/19061.ipynb). The code now successfully runs on GPU without taking longer duration.  Observe the result in screenshot. 

<img width="2132" height="1062" alt="Image" src="https://github.com/user-attachments/assets/ea4ce5cf-f1a1-4d67-bab9-47b6d69d99a8" />
 

Thank You

**Comment by odinsbane** [comment]
Wow, nice digging. If I use the version you provided on colab, it finishes in a few seconds. 

If I run it without the `jit_compile=False` it again takes more time. Though now it only takes 2 minutes. Even letting it run a couple epochs, the version with `jit_compile=True` goes from 130s down. to 30s. When `jit_compile=False` it goes from 3s down to 40ms.

I don't know if this should be closed because it still seems like an issue. Should it be re-titled or, or maybe closed and a new issue created.
